### PR TITLE
Support broadcast prepareReq by tree when using RPBFT

### PR DESCRIPTION
**Pull request by cyjseagull**
Support broadcast prepareReq by tree when using RPBFT

### CODE REVIEW

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ libconsensus/pbft/PBFTEngine.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fisco/FISCO-BCOS/pullRequest?prid=4808613&bid=15996470)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ libconsensus/pbft/PBFTEngine.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fisco/FISCO-BCOS/pullRequest?prid=4808613&bid=15996470)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ libconsensus/pbft/PBFTEngine.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fisco/FISCO-BCOS/pullRequest?prid=4808613&bid=15996470)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ libconsensus/pbft/PBFTEngine.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fisco/FISCO-BCOS/pullRequest?prid=4808613&bid=15996470)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ libconsensus/pbft/PBFTEngine.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fisco/FISCO-BCOS/pullRequest?prid=4808613&bid=15996470)

File: libconsensus/rotating_pbft/RotatingPBFTEngine.h [hunk 1]
```diff
@@ -22,6 +22,7 @@
  */
 #pragma once
 #include <libconsensus/pbft/PBFTEngine.h>
+#include <libsync/TreeTopology.h>
```
Comment (wheatli) [inline libconsensus/rotating_pbft/RotatingPBFTEngine.h]: DO NOT depend on libsync, plz move TreeTopology.h to lower layer

File: libledger/LedgerParam.h [hunk 2]
```diff
@@ -68,6 +68,9 @@ struct ConsensusParam
     // enable optimize ttl or not
     bool enableTTLOptimize;
     bool enablePrepareWithTxsHash;
+
+    bool broadcastPrepareByTree;
```
Comment (wheatli) [inline libledger/LedgerParam.h]: Make sure this option is only useable for rPBFT

File: libconsensus/rotating_pbft/RotatingPBFTEngine.h (hunk 1 above)
Comment (JimmyShi22) [reply libconsensus/rotating_pbft/RotatingPBFTEngine.h]: Maybe we can put it in `libdevcore`.

Comment (cyjseagull) [reply libconsensus/rotating_pbft/RotatingPBFTEngine.h]: > Maybe we can put it in `libdevcore`.

Yes, you're right, I will move `TreeTopology.h` to `devcore`

File: libledger/LedgerParam.h (hunk 2 above)
Comment (cyjseagull) [reply libledger/LedgerParam.h]: The `broadcastPrepareByTree` configuration is only used after init `RotatingPBFTEngine`, so this option is only used by `rPBFT`.

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ libconsensus/pbft/PBFTEngine.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fisco/FISCO-BCOS/pullRequest?prid=4808613&bid=15996470)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ libconsensus/pbft/PBFTEngine.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fisco/FISCO-BCOS/pullRequest?prid=4808613&bid=15996470)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ libconsensus/pbft/PBFTEngine.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fisco/FISCO-BCOS/pullRequest?prid=4808613&bid=15996470)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ libconsensus/pbft/PBFTEngine.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fisco/FISCO-BCOS/pullRequest?prid=4808613&bid=15996470)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ libconsensus/pbft/PBFTEngine.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fisco/FISCO-BCOS/pullRequest?prid=4808613&bid=15996470)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ libconsensus/pbft/PBFTEngine.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fisco/FISCO-BCOS/pullRequest?prid=4808613&bid=15996470)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ libconsensus/pbft/PBFTEngine.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fisco/FISCO-BCOS/pullRequest?prid=4808613&bid=15996470)

**Pull request merged**